### DISCUSSION TITLE
[V5] How to validate empty string but don't allow undefined and null?

### DISCUSSION THREAD

**Discussion by lu7766lu7766** [adonisjs/core#4112] [Help]
Validator schema:
```
name: schema.string.optional(),
```
Now I send { name: null } or { name: "" } or { name: undefined } , everything i requrest will pass, then

```
name: schema.string(),
```
Now { name: null } or { name: "" } or { name: undefined } all throw Exception

I want name is required, but empty string is OK. like...
{name: ""} OK
{name: "hello"} OK
{name: null} whatever
{name: undefined} Exception
{} Exception

**Comment by WeyIin** [comment]
Check this: 
https://docs.adonisjs.com/guides/validator/custom-rules

create custom rule and register it.

In the rule you want to check for nulls or undefined and throw a validation error, everything else pass.

e.g.

```javascript
validator.rule('notNullOrUndefined', (value, [field], options) => {
	if (value === null || value === undefined) {
		options.errorReporter.report(
			options.pointer,
			'notNullOrUndefined',
			'notNullOrUndefined validation failed',
			options.arrayExpressionPointer,
			{ field}
		)
	}
})
```

**Reply by lu7766lu7766** [reply]
but if i set the schema
```
name: schema.string.optional([rules.notNullOrUndefined()])
```
i send {} it will pass
actually i hope name is required and can be empty

**Reply by lu7766lu7766** [reply]
i find the way work on it
```
// set schema optional and add a custom rule
schema.string.optional([rules.stringAllowEmpty()])
// new rule
validator.rule(
  "stringAllowEmpty",
  (value, _, options) => {
    if (typeof value !== "string") {
      options.errorReporter.report(options.pointer, "stringAllowEmpty", "stringAllowEmpty validation failed", options.arrayExpressionPointer)
    }
  },
  // conpileFn let request required
  () => ({
    allowUndefineds: true,
  })
)
```
